### PR TITLE
Add GA4 tracking to local transactions

### PR DESCRIPTION
**Pull request by AshGDS**
⚠️ This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

Hi @andysellick, would you be able to approve this if all is OK? Thanks :+1:

## What

Adds the following GA4 events to the `local_transaction` flow (https://www.gov.uk/rubbish-collection-day)
- form_submit when clicking the "Find" button
- form_error when submitting an invalid postcode
- form_complete when you land on a results page 
- information_click if you click the result link

## Why
https://trello.com/c/2qmN8Wjg/546-form-events-formsubmit-formerror-formcomplete-and-informationclick-on-localtransaction-format

- This relies on this change in `govuk_publishing_components` to make the `form_submit` work, so the following PR should be deployed first: https://github.com/alphagov/govuk_publishing_components/pull/3409

## How to test it

1. Run `frontend` locally. If the publishing components PR is not deployed, run `frontend` with a local `static`, and have `static` point to local publishing components. Then make sure your publishing components is on the branch `ga4-local-transaction-changes`
2. Instead of setting up the APIs required for local transactions, we're going to fake the result. To do this open the file `app/controllers/local_transaction_controller.rb` in your editor.
3. Change the `fetch_postcode` function to this:
```Ruby
  def fetch_location(postcode)
    if postcode.present?
      local_custodian_codes = [5990] # hardcoded list of custodian codes
    end
    LocationsApiPostcodeResponse.new(postcode, local_custodian_codes, nil)
  end
```
4. Change the `authority_results` function to this
```Ruby
  def authority_results
    @authority_results = {
        "local_authorities" => [ 
          "name" => 'Example Council',
          "homepage_url" => "http://example.com",
          "country_name" => "England",
          "tier" => "district",
          "slug" => "westminster",
        ]
    }
  end
```
5. Change the `interaction_details` function to this
```Ruby
  def interaction_details
    {
      "local_authority" => {
        "name" => "Hello World",
        "homepage_url" => "http://westminster.example.com",
        "country_name" => "England",
      },
      "local_interaction" => { "url" => "http://westminster.example.com" },
    }
  end
```
6. Now if you go to http://127.0.0.1:3005/rubbish-collection-day/ and submit the form, it should redirect to a results page with our faked data.

### CODE REVIEW

File: app/views/local_transaction/results.html.erb [hunk 1]
```diff
@@ -8,10 +8,18 @@
...
        data-track-action="postcodeResultShown"
-       data-track-label="<%= @local_authority.name %>">
+       data-track-label="<%= @local_authority.name %>"
+       data-ga4-auto="<%= {
+        event_name: "form_complete",
+        type: "local transaction",
+        text: Nokogiri::HTML(t('formats.local_transaction.matched_postcode_html', local_authority: @local_authority.name, locale: :en)).text,
+        action: "complete",
+        tool_name: publication.title
+
```
Comment (andysellick) [inline app/views/local_transaction/results.html.erb:21]: Nitpick: unneeded empty line?

File: app/views/local_transaction/results.html.erb [hunk 2]
```diff
@@ -26,7 +26,16 @@
...
+      <p id="get-started" class="get-started group"
+      data-module="ga4-link-tracker"
+      data-ga4-track-links-only
+      data-ga4-set-indexes
+      data-ga4-link="<%= {
+        "event_name": "information_click",
+        "type": "local transaction",
+        "tool_name": publication.title,
+        "action": "information click"
+      }.to_json %>">
```
Comment (andysellick) [inline app/views/local_transaction/results.html.erb:38]: Nitpick: I think lines 30 to 38 should be indented.

File: test/integration/local_transactions_test.rb [hunk 3]
```diff
@@ -117,6 +117,17 @@ class LocalTransactionsTest < ActionDispatch::IntegrationTest
...
+      should "add the GA4 link tracker around the result link (information_click)" do
+        data_module = page.find("#get-started")["data-module"]
+        expected_data_module = "ga4-link-tracker"
+
+        ga4_link_attribute = page.find("#get-started")["data-ga4-link"]
+        ga4_expected_object = "{\"event_name\":\"information_click\",\"type\":\"local transaction\",\"tool_name\":\"Pay your bear tax\",\"action\":\"information click\"}"
+
+        assert_equal expected_data_module, data_module
+        assert_equal ga4_expected_object, ga4_link_attribute
+      end
```
Comment (andysellick) [inline test/integration/local_transactions_test.rb:140]: Could some of these commits that add a test be merged with their appropriate code change commit?

**Review by andysellick — APPROVED** [review]
Looks good, just a few minor nitpicks.

**Comment by andysellick** [comment]
Also great PR description and details, made reviewing this very easy, good job @AshGDS

**Comment by AshGDS** [comment]
Thanks @andysellick - I've made all the changes :+1: Yeah we used to use these types of commits (atomic commits) at my previous place, which works well for PRs!

**Pull request merged**